crates/miden-lib/asm/miden/contracts/auth/multisig.masm

```
@@ -0,0 +1,292 @@
+
+        movup.4 movdn.8 swapw dupw movdnw.2
+        # => [MESSAGE, OWNER_PUB_KEY, MESSAGE, i-1]
+
+        hmerge
+        # => [SIG_KEY, MESSAGE, i-1]
+
+        adv.has_mapkey
+        # => [SIG_KEY, MESSAGE, i-1]
+
+        adv_push.1
+        # => [has_signature, SIG_KEY, MESSAGE, i-1]
+
+        # if h(msg, pk) => msg exists in AdviceMap check the signature
+        if.true
+            dropw dupw
+            # => [MSG, MSG, i-1]
+
+            # Verify the signature against the public key and message.
+            # -----------------------------------------------------------------------------------------
+
+            padw mem_loadw.TEMP_PK
+            # => [PK, MSG, MSG, i-1]
+
+            swapw
+            # => [MSG, PK, MSG, i-1]
+
+            # Emit the authentication request event that pushes a signature for the message to the advice stack.
+            emit.AUTH_REQUEST
+
+            swapw
+            # => [PK, MSG, MSG, i-1]
+
+            exec.rpo_falcon512::verify
+            # => [MSG, i-1]
+
+            mem_load.SUCCESSFUL_VERIFICATIONS
+            add.1
+            mem_store.SUCCESSFUL_VERIFICATIONS
+            # => [MSG, i-1]
+
+        else
+            dropw
+            # => [MSG, i-1]
+        end
+        # => [MSG, i-1]
+
+        movup.4
+        # => [i-1, MSG]
+
+        dup neq.0
+        # => [should_continue, i-1, MSG]
+    end
+    # => [i-1, MSG]
+
+    # Check number of successful signature verifications
+    # -----------------------------------------------------------------------------------------
+
+    drop
+    # => [MSG]
+
+    mem_load.SUCCESSFUL_VERIFICATIONS
+    # => [num_verifications, MSG]
+
+    mem_load.THRESHOLD
+    # => [threshold, num_verifications, MSG]
+
+    gte not
+    # => [is_unauthorized, MSG]
+
+    # If signatures are non-existent the tx will fail here.
+    if.true
+        emit.UNAUTHORIZED_EVENT
+    end
+    # => [MSG]
```

This would only fail if the host is honest and decides to fail the transaction, but a malicious host might not do that and continue. So we have to add a `push.1 assertz.err="insufficient number of signatures"` or something after `emit` regardless to fail explicitly.